File: src/main/java/toby/command/commands/fetch/DbdRandomKillerCommand.java
package toby.command.commands.fetch;

import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import toby.command.CommandContext;
import toby.command.ICommand;
import toby.command.commands.misc.RandomCommand;
import toby.helpers.Cache;
import toby.helpers.WikiFetcher;
import toby.jpa.dto.UserDto;

import java.io.IOException;
import java.util.List;

public class DbdRandomKillerCommand implements IFetchCommand {

    private static final String dbdWebUrl = "https://deadbydaylight.fandom.com/wiki/Killers";
    public static final String cacheName = "dbdKillers";
    public static final String className = "mw-content-ltr";
    public static final String cssQuery = "div";
    private final Cache cache;

    public DbdRandomKillerCommand(Cache loadingCache) {
        this.cache = loadingCache;
    }

    @Override
    @SuppressWarnings("unchecked")
    public void handle(CommandContext ctx, UserDto requestingUserDto, Integer deleteDelay) {
        ICommand.deleteAfter(ctx.getEvent().getHook(), deleteDelay);
        final SlashCommandInteractionEvent event = ctx.getEvent();
        event.deferReply().queue();
        try {
            WikiFetcher wikiFetcher = new WikiFetcher(cache);
            List<String> dbdKillers = wikiFetcher.fetchFromWiki(cacheName, dbdWebUrl, className, cssQuery);
            event.replyFormat(RandomCommand.getRandomElement(dbdKillers)).queue(message -> ICommand.deleteAfter(message, deleteDelay));

        } catch (IOException ignored) {
            event.replyFormat("Huh, the website I pull data from must have returned something unexpected.").queue(message -> ICommand.deleteAfter(message, deleteDelay));
        }
    }

    @Override
    public String getName() {
        return "dbd";
    }

    @Override
    public String getDescription() {
        return "return a random dead by daylight killer";
    }

}

File: src/main/kotlin/toby/command/commands/fetch/DbdRandomKillerCommand.kt
package toby.command.commands.fetch

import toby.command.CommandContext
import toby.command.ICommand.Companion.invokeDeleteOnMessageResponse
import toby.command.commands.misc.RandomCommand
import toby.helpers.Cache
import toby.helpers.WikiFetcher
import toby.jpa.dto.UserDto
import java.io.IOException

class DbdRandomKillerCommand(private val cache: Cache) : IFetchCommand {
    override fun handle(ctx: CommandContext, requestingUserDto: UserDto, deleteDelay: Int?) {
        val event = ctx.event
        event.deferReply().queue()
        try {
            val wikiFetcher = WikiFetcher(cache)
            val dbdKillers = wikiFetcher.fetchFromWiki(cacheName, dbdWebUrl, className, cssQuery)
            event.hook.sendMessageFormat(RandomCommand.getRandomElement(dbdKillers)).queue(invokeDeleteOnMessageResponse(deleteDelay!!))
        } catch (ignored: IOException) {
            event.hook.sendMessageFormat("Huh, the website I pull data from must have returned something unexpected.").queue(invokeDeleteOnMessageResponse(deleteDelay!!))
        }
    }

    override val name: String
        get() = "dbd-killer"
    override val description: String
        get() = "return a random dead by daylight killer"

    companion object {
        private const val dbdWebUrl = "https://deadbydaylight.fandom.com/wiki/Killers"
        const val cacheName = "dbdKillers"
        const val className = "mw-content-ltr"
        const val cssQuery = "div"
    }
}
